### PR TITLE
LRU Cache for Boxes/KVs

### PR DESCRIPTION
**Pull request by algoidurovic**
## Summary

This KV cache is intended to nearly exactly replicate the resources cache, in lieu of generics. Included are also the box database benchmark tests that were the original motivation for implementing a cache. One of the size limits of the cache (`baseKVPendingAccountsBufferSize = 50,000`) is informed by the maximum box size (~32KB) and the `baseResourcesPendingAccountsBufferSize = 100,000` where the maximum resource size is ~16KB. The other limitation on cache size is `len(au.kvStore)`: the number of KV modifications in the state deltas. While the theoretical maximum memory usage is quite high, this is an existing problem with the current cache implementations and should be addressed in another PR.

## Test Plan

The cache and underlying data structures are tested in unit tests.

### CODE REVIEW

File: ledger/accountdb.go
Comment (algoidurovic) [inline ledger/accountdb.go:2570]: I just renamed and moved this but I don't mind undoing it if there's an objection.

File: ledger/acctupdates.go
Comment (algoidurovic) [inline ledger/acctupdates.go:390]: Error handling has been slightly changed here, but seems innocent enough to me.

File: ledger/acctupdates.go
Comment (ahangsu) [inline ledger/acctupdates.go:394]: question: even if we read it in cache, we also rerank the key to be the latest touched, right?

File: ledger/acctupdates.go
Comment (ahangsu) [inline ledger/acctupdates.go:409]: question: does that mean either one of following case is true here:
- `err != nil` and `persistedData.value == nil`
- `err == nil` and `persistedData.value != nil`

If not, then I would wonder if we covered all cases...

File: ledger/acctupdates.go
Comment (ahangsu) [inline ledger/acctupdates.go:73]: this parameter looks interesting, is that based on some estimation of byte size of storage?

File: ledger/lruboxes.go
Comment (ahangsu) [inline ledger/lruboxes.go:96]: question: if we write to a channel, is this the only way to do? or is there any special motivation behind such specific form of code? I wonder if JJ @jannotti knows anything tho...

**Review by ahangsu — COMMENTED** [review]
just some random questions, the implementation looks following previous implementation patterns, should be legit

File: ledger/lruboxes.go
Comment (jannotti) [reply ledger/lruboxes.go:96]: This is to avoid blocking, and drop the message instead if the channel buffer is full.
https://gobyexample.com/non-blocking-channel-operations

File: ledger/acctupdates.go
Comment (ahangsu) [reply ledger/acctupdates.go:73]: nvm, I just read the estimation in description...

File: ledger/persistedboxes_list.go
Comment (michaeldiamant) [inline ledger/persistedboxes_list.go:20]: @algoidurovic Is the TODO still relevant?

I leave a _suggestion_ to update (what appears to be) a stale comment reference.

```suggestion
must initiate with newPersistedBoxList.
```

File: ledger/lruboxes.go
Comment (michaeldiamant) [inline ledger/lruboxes.go:23]: @algoidurovic Sanity checking:  Is it by-design to refer to `cachedResourceData` here?  

If so, I'd appreciate a quick note to build my intuition.  Based on other cache implementations, I leave a _suggestion_.

```suggestion
//msgp:ignore cachedBoxData
```

File: ledger/acctupdates.go
Comment (michaeldiamant) [inline ledger/acctupdates.go:418]: @algoidurovic Leaving note on unrelated line because impacted line is _not_ in diff.

In `lookupKeysByPrefix`on line 526, it appears there's a placeholder to call the LRU cache.  Double checking: Does the PR deliberately make no change?

I spotted the concern by inspecting usage of `lruboxes.go::read` and considering other possible call sites.

File: ledger/lruboxes_test.go
Comment (michaeldiamant) [inline ledger/lruboxes_test.go:208]: @algoidurovic What does the TODO signify here?

I'm _guessing_ it's to replace the _account_ constants with _box_ constants.

File: ledger/accountdb.go
Comment (michaeldiamant) [reply ledger/accountdb.go:2570]: @algoidurovic I _think_ the rename touches on a central question:  What is box-specific about the cache?

I realize the question prompts a potentially large renaming exercise, but I'm wondering if the cache ought to be _kv_ themed rather than _box_ themed.

File: ledger/lruboxes_test.go
Comment (michaeldiamant) [reply ledger/lruboxes_test.go:208]: Sorry - replace _account_ with _resources_ in my prior note.

File: ledger/acctupdates.go
Comment (michaeldiamant) [inline ledger/acctupdates.go:71]: @algoidurovic and reviewers:  Do we agree the reference to _accounts_ can be removed?

I realize the convention was copied from the _resources_ constant.  I don't see how either set of constants is _account_ related.

File: ledger/acctupdates.go
Comment (michaeldiamant) [inline ledger/acctupdates.go:418]: @algoidurovic Leaving another note on an unrelated line because impacted areas are not in the diff.

Double checking my understanding:  Is it accurate to expect `au.baseBoxes` updates in `postCommit`?

I see `au.baseResources` updates happening based on return types from `makeCompactResourceDeltas`.

File: ledger/accountdb.go
Comment (jannotti) [reply ledger/accountdb.go:2570]: yes, I think it should.  Anything below the ledger tries to talk about kv rather than about boxes. But I don't think it changes much code.

File: ledger/acctupdates.go
Comment (algoidurovic) [reply ledger/acctupdates.go:394]: Yes, that's the purpose of the code you highlighted. It's a little more complicated though because the writes are first buffered than flushed later. The higher level goal is the same.

File: ledger/acctupdates.go
Comment (algoidurovic) [reply ledger/acctupdates.go:71]: I agree with you but kept it just in case I missed something. Personally, I have no problem removing the comment.

File: ledger/acctupdates.go
Comment (algoidurovic) [reply ledger/acctupdates.go:409]: That seems correct to me. I think these semantics are a little unusual but they're consistent with resource lookup and cache entry promotion. It seems like the goal is to make sure the cache is updated even if the err != nil.

File: ledger/acctupdates.go
Comment (algoidurovic) [reply ledger/acctupdates.go:418]: I deliberately didn't use the cache there because it isn't set up to make queries on box name prefixes. Also, `lookupKeysByPrefix` is only called by the API server afaik so it's probably not as important to speed up reads with a cache.

File: ledger/acctupdates.go
Comment (michaeldiamant) [reply ledger/acctupdates.go:71]: @algoidurovic Sorry - I'm suggesting to rename related variables without referencing _accounts_, rather than rewording the comment.

Comment (algoidurovic) [reply ledger/acctupdates.go:71]: After gaining for familiarity with the code, I believe the reference to accounts can be removed in the variable name. Boxes are totally decoupled from accounts at this level, unlike resources, so the convention doesn't need to carry over.

File: ledger/acctupdates.go
Comment (algoidurovic) [reply ledger/acctupdates.go:418]: Hopefully, this has been addressed in the latest commits. The new test in `accountupdates_test.go` does fail on the code when your comment was written, and now is passing after implementation gaps were filled.

File: ledger/persistedboxes_list.go
Comment (algoidurovic) [reply ledger/persistedboxes_list.go:20]: I wasn't sure what the purpose of this this TODO was. The resource cache had the same comment so I copied it over. Maybe @algorandskiy has some input?

File: ledger/lruboxes.go
Comment (algoidurovic) [reply ledger/lruboxes.go:23]: fixed.

File: ledger/accountdb_test.go
Comment (jannotti) [inline ledger/accountdb_test.go:322]:
```suggestion
		require.Empty(t, updatedBoxes)
```

File: ledger/accountdb_test.go
Comment (jannotti) [inline ledger/accountdb_test.go:466]:
```suggestion
			require.Empty(t, updatedBoxes)
```

File: ledger/accountdb_test.go
Comment (jannotti) [inline ledger/accountdb_test.go:2897]: These could all use `require.Len`

File: ledger/accountdb_test.go
Comment (jannotti) [inline ledger/accountdb_test.go:3024]: Instead of Stop/Start, you can just use b.ResetTimer()

File: ledger/acctupdates.go
Comment (jannotti) [reply ledger/acctupdates.go:409]: I think we want to writePending() whether the `value` is nil or not.  That will, effectively, mean we're caching negative responses, which seems worthwhile.

The other caches don't have a way to store a negative result, but we do (nil pointer to string).

I would change my mind if this leads to any changes in the cache itself.

File: ledger/acctupdates.go
Comment (jannotti) [reply ledger/acctupdates.go:418]: I think this is the right decision.  We'd need a new datastructure to support range queries, and we'd have to think hard about cache coherence.  Let's skip it.  However, let's note that this is a point in favor of giving the caching job to sqlite (bigger cache? bigger pages?) as it would probably help there.

File: ledger/acctupdates_test.go
Comment (jannotti) [inline ledger/acctupdates_test.go:1364]: Why are you setting these? I don't think they do what you think.  See the most recent merge to feature/avm-box, where I had to change this.  You (probably) want conf.MaxAcctLookback.

File: ledger/acctupdates_test.go
Comment (jannotti) [inline ledger/acctupdates_test.go:1414]: Yeah, I think `protoParams.MaxBalLookback` should be `conf.MaxAcctLookback`

File: ledger/acctupdates_test.go
Comment (jannotti) [inline ledger/acctupdates_test.go:1415]: same, again

File: ledger/acctupdates.go
Comment (michaeldiamant) [reply ledger/acctupdates.go]: @algoidurovic Resolving - Thanks for the rename in https://github.com/algorand/go-algorand/pull/4275/commits/f40283e66555baa0835b5aca70b1782c87493c3a.

File: ledger/lruboxes_test.go
Comment (michaeldiamant) [reply ledger/lruboxes_test.go]: @algoidurovic Resolving - Thanks for https://github.com/algorand/go-algorand/pull/4275/commits/abaf6ca575a9589e3791ebed53cddf8ccc5f62a3.

File: ledger/persistedboxes_list.go
Comment (michaeldiamant) [reply ledger/persistedboxes_list.go:20]: @algoidurovic Where do you see the TODO in resource cache?  I don't see a corresponding TODO in `persistedresources_list.go`.

File: ledger/acctupdates.go
Comment (michaeldiamant) [reply ledger/acctupdates.go:418]: @algoidurovic @jannotti I'm on-board with _not_ supporting the prefix use case.  

Prior to closing the thread, can an inline comment be added to `lookupKeysByPrefix` that explains _why_ there's no LRU cache usage?

File: ledger/accountdb.go
Comment (michaeldiamant) [inline ledger/accountdb.go:3640]: @algoidurovic Does the following _suggestion_ seem plausible to you?

I imagine (have not researched closely) it _hardly_ impacts performance, though it seems we can easily + accurately pre-allocate `updatedBoxes` here.

```suggestion
	updatedBoxes = make(map[string]persistedBoxData, len(kvPairs))
```

File: ledger/acctupdates.go
Comment (michaeldiamant) [inline ledger/acctupdates.go:73]: @algoidurovic Can the inline comment be expanded (or a new one added) to summarize _why_ we chose the configured value?

I realize the value is liable to change prior to merge.  Consider the note a visual reminder for when we're ready to finalize the value.

File: ledger/acctupdates_test.go
Comment (michaeldiamant) [inline ledger/acctupdates_test.go:1357]: @algoidurovic Minor and optional:  Is there a by-design reason _not_ to run the test in parallel?

Running `gotestsum -- ./ledger` locally several times works without issue.  Cursory inspection suggests to me the test does not rely on global state changes.

```suggestion
	partitiontest.PartitionTest(t)
        t.Parallel()

```

File: ledger/acctupdates_test.go
Comment (michaeldiamant) [inline ledger/acctupdates_test.go:1421]: @algoidurovic Can you help me understand the significance of the if-statement?

When I remove these statements wrapping `kvMods` check, the test still passes.  It makes me think the if-statement is extraneous.

File: ledger/acctupdates_test.go
Comment (michaeldiamant) [inline ledger/acctupdates_test.go:1377]: @algoidurovic As a heads up, I created a tangentially related PR:  https://github.com/algorand/go-algorand/pull/4353.  If https://github.com/algorand/go-algorand/pull/4353 is merged ahead of this PR, then a minor change will be needed.

Comment (michaeldiamant) [reply ledger/acctupdates_test.go:1377]: @algoidurovic FYI - #4353 is merged.  When you merge master in, you'll have minor compile-time errors to resolve.  Let me know if you'd like to pair.

File: ledger/persistedboxes_list.go
Comment (algoidurovic) [reply ledger/persistedboxes_list.go:20]: Sorry, the TODO wasn't in `persistedresources_list.go`, the comment itself was. The TODO was added by me to look into if this comment would also apply to `newPersistedBoxList`. On second inspection, it looks like the original comment in the resources file was erroneously copied over from the account file. I'll change that as well.

File: ledger/accountdb.go
Comment (algoidurovic) [reply ledger/accountdb.go:3640]: Seems reasonable, I'll add it.

File: ledger/acctupdates.go
Comment (algoidurovic) [reply ledger/acctupdates.go:409]: I agree, but I think we still need to guard against pushing a write to the cache when an err != nil, right?

File: ledger/acctupdates_test.go
Comment (algoidurovic) [reply ledger/acctupdates_test.go:1421]: I think it technically is extraneous. Boxes are only being manipulated for `boxCnt/boxesPerBlock` blocks, and the remaining iterations of the loop are meant to verify the cache is in the right state after all the blocks have been committed. I'll refactor this into two loops to make it more clear.

Comment (algoidurovic) [reply ledger/acctupdates_test.go:1421]: Actually, I take it back. The body of the loop can't be divided up easily. I'll just add more comments to describe what's going on.

File: ledger/acctupdates.go
Comment (algoidurovic) [reply ledger/acctupdates.go:420]: Added a comment.

Comment (michaeldiamant) [reply ledger/acctupdates.go:420]: @algoidurovic Resolving - thanks for the inline comment.

Comment (michaeldiamant) [reply ledger/acctupdates.go:420]: @algoidurovic New test in `accountupdates_test.go` covers it from my end - thanks + resolving.

File: ledger/accountdb_test.go
Comment (michaeldiamant) [inline ledger/accountdb_test.go:314]: @algoidurovic Minor:  As part of renaming from Boxes-to-KV, can you make these renames?
* Rename `updatedBoxes` to `updatedKVs` to match the method being tested.
* Rename PR title to include a _KV_ mention.  
  * I periodically search PR history and having the associated keywords simplifies historical searches.
  * If you'd prefer, I'm good with retaining a Box reference in the PR title too.

File: ledger/acctupdates_test.go
Comment (michaeldiamant) [reply ledger/acctupdates_test.go]: @algoidurovic Resolving - Thanks for the inline comment + test implementation change here.

File: ledger/accountdb.go
Comment (michaeldiamant) [reply ledger/accountdb.go:2570]: @algoidurovic From my perspective, based on renames to _KV_, the thread is now resolved.  Since I didn't open it, I'm leaving it open for you to resolve.

File: ledger/acctupdates.go
Comment (michaeldiamant) [reply ledger/acctupdates.go]: > we still need to guard against pushing a write to the cache when an err != nil, right?

Chiming in as a reviewer - I _think_ the desired behavior is implemented though I leave it to original commenters to verify + resolve.
* As of https://github.com/algorand/go-algorand/pull/4275/commits/25946b42b8c511286a9341ebb42657538adcfe30, cache is _not_ updated when `err != nil`.  Otherwise, it's updated.  
* I think _not_ updating when `err != nil` _is_ the desired behavior.

File: ledger/accountdb_test.go
Comment (michaeldiamant) [reply ledger/accountdb_test.go:314]: Additionally, any updates to the PR title ought to be reflected into the PR description.

File: ledger/acctupdates.go
Comment (ahangsu) [reply ledger/acctupdates.go]: yes @michaeldiamant is correct, this thread is resolving

File: ledger/acctupdates_test.go
Comment (algoidurovic) [reply ledger/acctupdates_test.go:1374]: I think this can be resolved now.

**Review by michaeldiamant — APPROVED** [review]
@algoidurovic Thanks for your effort to implement + test a KV-specific cache implementation!

Approving with the understanding that we'll finalize configuration of `baseKVPendingBufferSize` out-of-band to the PR.

File: ledger/accountdb.go
Comment (jannotti) [inline ledger/accountdb.go:2603]: This became `persistedKVData`?  I guess I don't object too much, but it is not a Key and Value, it's only a value, which is why I called it this. I admit "persistedValue" ends up sounds pretty generic.

File: ledger/accountdb.go
Comment (jannotti) [inline ledger/accountdb.go:4821]: Since they are small objects, I might not use pointers for the receiver or the arg.

File: ledger/acctupdates.go
Comment (jannotti) [inline ledger/acctupdates.go:218]: I'd call it either `baseKVs` or `baseKVStore`, as it is a bunch of Key/Values. (right?)

File: ledger/acctupdates.go
Comment (jannotti) [inline ledger/acctupdates.go:409]: Pavel recently made a PR to change the error handling on the similar functions that would be good to copy here. 

#4425

File: ledger/acctupdates.go
Comment (jannotti) [inline ledger/acctupdates.go:1338]: Why is this PR changing lookupResource?

**Review by jannotti — COMMENTED** [review]
Some feedback, I'm going to go pull this down locally and diff the lru now, since github review  isn't good for that.

File: ledger/acctupdates.go
Comment (jannotti) [inline ledger/acctupdates.go:1888]: What can we do to get coverage of this? It seems important.

File: ledger/accountdb.go
Comment (jannotti) [reply ledger/accountdb.go:2603]: Looking further into lrukv.go, it seems like we ought to keep the name you changed it to, but we ought to add `key` to it.  For accounts, `persistedAccountData` has the address AND the account data, I thing the analogy here would be for us to hold the key AND value.

Comment (algoidurovic) [reply ledger/accountdb.go:2603]: Makes sense to me, I'll modify.

File: ledger/acctupdates.go
Comment (algoidurovic) [reply ledger/acctupdates.go:1338]: Good catch, I don't know how that got in there.

File: ledger/acctupdates.go
Comment (algoidurovic) [reply ledger/acctupdates.go:1888]: My brief research indicates that this method is only called during specific db migrations triggered in `trackerdb.go`'s `runMigration` function. ie only when the ledger is being reloaded during an upgrade of the db schema version from 2 to 3. Seems difficult to test considering we're on version 8 and there aren't any tests I can find for the 2->3 migration.

Comment (jannotti) [reply ledger/acctupdates.go:1888]: Ok, I was thinking this happened periodically.

File: ledger/accountdb.go
Comment (algoidurovic) [reply ledger/accountdb.go:2603]: On second thought, this makes me a little nervous because I modeled the kv cache after the resource cache, which decouples the address from the persisted data object. Not sure why it works that way or if it should be changed.

Comment (jannotti) [reply ledger/accountdb.go:2603]: Agree! Thanks.

File: ledger/acctupdates.go
Comment (jannotti) [inline ledger/acctupdates.go:417]: Now that `err` is checked earlier, let's make this explicitly `nil` so it is obvious it's the good path.

Comment (algoidurovic) [reply ledger/acctupdates.go]: got it, fixed.

**Review by jannotti — APPROVED** [review]

**Pull request merged**
